### PR TITLE
RSDK-5368 Add more safety and inline docs about logical clocks in resource graph

### PR DESCRIPTION
**Pull request by benjirewis**
RSDK-5368

The “logical clock” logic shared between local robot, the resource graph and graph nodes represents a somewhat complex system used to determine when `updateWeakDependents` should be run in the local robot. There is both a lack of comments explaining this logic and some inaccurate statements in the existing comments and inline documentation. This PR adds more comments, attempts to fix the inaccuracies, and adds a couple "missing" locks.

### CODE REVIEW

File: resource/resource_graph.go [hunk 1]
```diff
@@ -37,8 +40,8 @@ func NewGraph() *Graph {
 	}
 }
 
-// LastUpdatedTime returns the last logical clock time a resource updated.
```
Comment (benjirewis) [inline resource/resource_graph.go:40]: Given that I'm eventually adding a `time.Time` `lastReconfigured` field to graph nodes, `LastUpdatedTime` _seems_ like it would return a `time.Time`. I think it's more accurate to call this `LogicalClock` as it simply returns the value of that atomic.

File: robot/impl/local_robot.go [hunk 2]
```diff
@@ -525,6 +527,10 @@ func (r *localRobot) getDependencies(
 	var needUpdate bool
 	for _, dep := range r.manager.resources.GetAllParentsOf(rName) {
 		if node, ok := r.manager.resources.Node(dep); ok {
+			// If any of the parents of the resource we're getting dependencies for
+			// has an updatedAt value that is "later" than the last value at which we
+			// ran updateWeakDependents, ensure that we run updateWeakDependents
+			// later in this method.
```
Comment (benjirewis) [inline robot/impl/local_robot.go:533]: It's unclear to me _why_ we do this check and run `updateWeakDependents` _here_. I guess if we find _some_ parent of _some_ resource has a "later" clock than the last time we ran `updateWeakDependents`, we `updateWeakDependents` later in this function. I don't understand why we can't just check the graph's logical clock with `LastUpdatedTime` (now called `LogicalClock`) and run `updateWeakDependents` if it's "later" than `lastWeakDependentsRound`. 

I'm hesitant to change this code, as it's pretty hard for me to answer the question "when do we update weak dependents?". I think it might be a lot, and, to be honest, I'm not even sure why we have all this logical clock logic _just_ to not run `updateWeakDependents` too much.

**Review by benjirewis — COMMENTED** [review]
I realize my inline comments and my GitHub comments are pretty wordy; I'm mostly trying to figure out how this "logical clock" system works and documenting my understanding along the way in case someone else needs to understand this code at some point. I suppose I'm wondering if my ramblings make sense 😆 .

File: robot/impl/local_robot.go (hunk 2 above)
Comment (stuqdog) [reply robot/impl/local_robot.go:533]: > I don't understand why we can't just check the graph's logical clock with LastUpdatedTime (now called LogicalClock) and run updateWeakDependents if it's "later" than lastWeakDependentsRound.

I may not be understanding fully, but it seems to me that this proposal (i.e. running `updateWeakDependents` based on a comparison of `LogicalClock` and `lastWeakDependentsRound`) is logically equivalent to the current rule, but cleaner? `UpdatedAt` should always be set to a value equal to `LogicalClock` at the time of call, so unless there's some other place that `logicalClock` is getting updated somewhere else I think the proposed change is just cleaner. But this is all a bit complicated so totally reasonable to leave it as-is or save for another PR.

**Review by stuqdog — APPROVED** [review]
This all looks reasonable to me! Thanks for this; will definitely make working through this code a bit clearer in the future 😅

File: resource/resource_graph.go (hunk 1 above)
Comment (benjirewis) [reply resource/resource_graph.go:40]: Great idea; switched to `CurrLogicalClockValue`.

File: robot/impl/local_robot.go (hunk 2 above)
Comment (benjirewis) [reply robot/impl/local_robot.go:533]: > I may not be understanding fully, but it seems to me that this proposal (i.e. running updateWeakDependents based on a comparison of LogicalClock and lastWeakDependentsRound) is logically equivalent to the current rule, but cleaner?

Yep! I _think_ they'd be logically equivalent. Let me mess around locally for a bit and see if anything breaks if I move this check to the top of `updateWeakDependents` and use `CurrLogicalClockValue`.

File: robot/impl/local_robot.go [hunk 3]
```diff
@@ -655,9 +649,12 @@ func (r *localRobot) newResource(
 }
 
 func (r *localRobot) updateWeakDependents(ctx context.Context) {
-	// track that we are current in resources up to the latest update time. This will
-	// be used to determine if this method should be called while completing a config.
-	r.lastWeakDependentsRound.Store(r.manager.resources.LastUpdatedTime())
+	// Only run updateWeakDependents when the resource graph's logical clock is
+	// now "later" than the value was when we last ran updateWeakDependents.
+	if r.lastWeakDependentsRound.Load() >= r.manager.resources.CurrLogicalClockValue() {
```
Comment (benjirewis) [inline robot/impl/local_robot.go:654]: I moved all the `lastWeakDependentsRound` logic into `updateWeakDependents`, so we just early return if the current logical clock value is equal to or less than the last time we ran `updateWeakDependents`. I _think_ everything still passes, and this reads as logically equivalent to me.

`if r.lastWeakDependentsRound.Load() >= r.manager.resources.CurrLogicalClockValue()` is _not_ a perfect logical inversion of the previous conditional `if r.lastWeakDependentsRound.Load() <= node.UpdatedAt()`; I used `>=` instead of `>`. I think the original conditional _should_ have been `if r.lastWeakDependentsRound.Load() < node.UpdatedAt()`: if the last time we ran `updateWeakDependents` and the current node have an equivalent value, there's no reason to think the graph has been relevantly updated. So, the inversion _should_ probably be `>=`. Let me know if you disagree.

**Review by benjirewis — COMMENTED** [review]
I'd love to mess around with trying to remove the "logical clock" logic entirely and answer the questions "when should we update weak dependents?" and "when do we update weak dependents?", but it's probably too much work for right now, and I'm worried I'm going to start breaking things I don't understand.

File: robot/impl/local_robot.go (hunk 3 above)
Comment (stuqdog) [reply robot/impl/local_robot.go:654]: That sounds right to me!

Comment (benjirewis) [reply robot/impl/local_robot.go:654]: This change caused [`TestSensorsServiceReconfigure`](https://github.com/viamrobotics/rdk/blob/main/robot/impl/robot_reconfigure_test.go#L2255) to fail (a sort of tangential test of the weak dependency system). 

Unfortunately, checking whether we should update weak dependents by comparing directly against the current graph logical clock value results in too many weak dependency updates (4 vs 1 for subtests within `TestSensorsServiceReconfigure`). The old conditional would only update weak dependents when _some_ resource had _some_ dependency with an `updateAt` value greater than or equal to the last time we ran `updateWeakDependents`; that's obviously slightly more conservative as it examines only pieces of the graph at a time.

If you're curious: the test failed because the sensors service, which is weakly dependent on movement sensors, still found movement sensors even after they were removed from the robot. I don't really understand how that error could result from _too many_ weak dep updates 🤷🏻 . I was looking into it but found that `updateWeakDependents` is sometimes [recursive](https://github.com/viamrobotics/rdk/blob/main/robot/impl/local_robot.go#L764) (that `getDependencies` can call `updateWeakDependents` again), so I gave up trying to understand this for now.

TL;DR I'm not gonna modify the check because I'm definitely breaking something. This PR will remain only comment + lock additions 🙏🏻 .

Comment (dgottlieb) [reply robot/impl/local_robot.go]: I'm in agreement that the old condition was wrong. Thanks for looking into why your changes weren't working as expected.

> I'd love to mess around with trying to remove the "logical clock" logic entirely and answer the questions "when should we update weak dependents?" and "when do we update weak dependents?", but it's probably too much work for right now, and I'm worried I'm going to start breaking things I don't understand.

I'd also like to explore swapping out the logical clock for something simpler in the future. Now that I see it's purpose, it definitely feels like more of a local "boolean didSomethingChange" variable.

And calling this a logical clock is a bit of a misnomer. It counts like one, but it's not really* being used to establish any happens-before relationships. The naming comes off as a bit intimidating and inaccessible.

**Review by dgottlieb — APPROVED** [review]

File: robot/impl/local_robot.go (hunk 3 above)
Comment (benjirewis) [reply robot/impl/local_robot.go]: > Now that I see it's purpose, it definitely feels like more of a local "boolean didSomethingChange" variable.

> And calling this a logical clock is a bit of a misnomer. It counts like one, but it's not really* being used to establish any happens-before relationships. The naming comes off as a bit intimidating and inaccessible.

Well said and agree.
